src/envoy/tcp/alpn_proxy/alpn_proxy.cc

```diff
@@ -0,0 +1,164 @@
+      filter_direction_(filter_direction),
+      stats_(generateStats(stat_prefix, scope)) {}
+
+Network::FilterStatus AlpnProxyFilter::onData(Buffer::Instance& data, bool) {
+  // No work needed if we have read connection protocol and its not the expected
+  // alpn protocol or if alpn protocol is found but we have yet not consumed
+  // metadata.
+  if ((connection_protocol_read_ && !alpn_protocol_found_) ||
+      (alpn_protocol_found_ && !done_on_data_)) {
+    return Network::FilterStatus::Continue;
+  }
+  if (read_callbacks_->connection().nextProtocol() != config_->protocol_) {
+    connection_protocol_read_ = true;
+    config_->stats().alpn_protocol_not_found_.inc();
+    return Network::FilterStatus::Continue;
+  }
+  // Write metadta first time protocol is found.
+  if (!alpn_protocol_found_) {
+    config_->stats().alpn_protocol_found_.inc();
+    alpn_protocol_found_ = true;
+    if (local_info_.node().has_metadata()) {
+      auto metadata_fields = local_info_.node().metadata().fields();
+      auto node_metadata = metadata_fields.find(config_->node_metadata_id_);
+      if (node_metadata != metadata_fields.end()) {
+        std::unique_ptr<::Envoy::Buffer::OwnedImpl> buf =
+            ConstructProxyHeaderData(node_metadata->second.struct_value());
+        write_callbacks_->injectWriteDataToFilterChain(*buf, false);
+
+        read_callbacks_->connection().streamInfo().setDynamicMetadata(
+            downstreamDynamicDataKey, node_metadata->second.struct_value());
+        config_->stats().metadata_added_.inc();
+      }
+    }
+  }
+
+  // Consume Metadata in Downstream Filter.
+  if (config_->filter_direction_ ==
+      envoy::tcp::alpnproxy::config::AlpnProxy_FilterDirection::
+          AlpnProxy_FilterDirection_DOWNSTREAM) {
+    config_->stats().downstream_filter_found_.inc();
+    const uint32_t initial_header_length = sizeof(AlpnProxyInitialHeader);
+    if (data.length() < initial_header_length) {
+      config_->stats().initial_header_not_found_.inc();
+      // Not enough data to read. Wait for it to come.
+      return Network::FilterStatus::StopIteration;
+    }
+    std::string initial_header_buf(initial_header_length, 0);
+    data.copyOut(0, initial_header_length, &initial_header_buf[0]);
+    const AlpnProxyInitialHeader* initial_header =
+        reinterpret_cast<const AlpnProxyInitialHeader*>(
+            initial_header_buf.c_str());
+    if (absl::gntohl(initial_header->magic) !=
+        AlpnProxyInitialHeader::magic_number) {
+      config_->stats().initial_header_not_found_.inc();
+      done_on_data_ = true;
+      return Network::FilterStatus::Continue;
+    }
+    const uint64_t proxy_header_length =
+        absl::gntohl(initial_header->data_size);
+    // Drain the initial header length bytes read.
+    data.drain(initial_header_length);
+    if (data.length() < proxy_header_length) {
+      // Not enough data to read. Wait for it to come.
+      return Network::FilterStatus::StopIteration;
```

Comment:
You already drained initial_header_length above, in the next iteration of onData when new data come, it will read first `initial_header_length` again but the data no longer contains real magic/length.

Reply:
Yeah, I was hoping to refactor this in next PR but just did in this PR only :)